contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/azkaban/GobblinDistcpJob.java

```diff
@@ -0,0 +1,274 @@
+                }
+            }
+
+            if (props.getBoolean(VoldemortBuildAndPushJob.PUSH_CDN_WRITTEN_BY_OTHER, false)) {
+                if (!o.implies(FsAction.WRITE)) {
+                    o = o.or(FsAction.WRITE);
+                    changed = true;
+                }
+            }
+        }
+
+        if (changed) {
+            FsPermission desiredPerm = new FsPermission(u, g, o);
+            fs.setPermission(path, desiredPerm);
+            if (!fs.getFileStatus(path).getPermission().equals(desiredPerm)) {
+                throw new RuntimeException("Failed to set permission for " + path + " from: " + perm + " to: " + desiredPerm);
+            }
+            info("for path " + path + ", permissions changed from: " + perm + " to: " + desiredPerm);
+        }
+    }
+
+    private void addPermissionsToParentDirs(FileSystem fs, String target, String pathPrefix) throws Exception {
+        String pathFromRoot = removeLeadingSlash(extractPathFromUrl(target));
+        String parent = "";
+
+        for (String seg: pathFromRoot.split("/")) {
+            parent = parent + "/" + seg;
+            Path parentPath = new Path(parent);
+
+            if (parent.length() <= pathPrefix.length()) {
+                checkPermission(fs, parentPath, false);
+            } else {
+                checkPermission(fs, parentPath, true);
+            }
+        }
+    }
+
+    private Map<String, String> buildCdnMap() throws Exception {
+        List<String> pairs = props.getList(VoldemortBuildAndPushJob.PUSH_CDN_CLUSTER);
+        Map<String, String> cdnMap = new HashMap<>(pairs.size());
+
+        for (String pair : pairs) {
+            if (!pair.contains("|")) {
+                throw new RuntimeException("Cannot find separator '|' in K-V pair \"" + pair + "\"");
+            }
+            String[] urls = pair.split("\\|");
+            if (urls.length != 2) {
+                throw new RuntimeException("More than one separator found in K-V pair \"" + pair + "\"");
+            }
+            cdnMap.put(removeTrailingSlash(urls[0].trim()), removeTrailingSlash(urls[1].trim()));
+        }
+        return cdnMap;
+    }
+
+    private String pickCDN() throws Exception {
+        String cdnCluster = buildCdnMap().get(removeTrailingSlash(destination));
+
+        if (cdnCluster == null) {
+            warn("Cannot find corresponding CDN! Will bypass CDN for push cluster " + destination);
+            return "";
+        }
+
+        if (cdnCluster.equals("null")) {
+            info("Will bypass CDN for push cluster " + destination);
+            return "";
+        }
+
+        if (!cdnCluster.matches(".*hdfs://.+:[0-9]{1,5}")) {
+            warn("Invalid URL format! Will bypass CDN for push cluster " + destination);
+            return "";
+        }
+        return cdnCluster;
+    }
+
+    private boolean prereqSatisfied(String cdnURL) throws Exception {
+        for (String namenode: props.getList("other_namenodes")) {
```

Comment:
"other_namenodes" appears several times, so could we use a constant string instead?